File: Python/rk4.py
from SimulationInitialization import zero, int2, int6


def rk4 (dydt, tspan, y0, m, y):
    f0 = [0.0, 0.0, 0.0, 0.0]
    f1 = [0.0, 0.0, 0.0, 0.0]
    f2 = [0.0, 0.0, 0.0, 0.0]
    f3 = [0.0, 0.0, 0.0, 0.0]
    u1 = [0.0, 0.0, 0.0, 0.0]
    u2 = [0.0, 0.0, 0.0, 0.0]
    u3 = [0.0, 0.0, 0.0, 0.0]
    dt = tspan[1] - tspan[0]
    t = tspan[0]

    dydt(t, y0, f0)

    for i in range(m):
        u1[i] = y0[i] + dt * f0[i] / int2
    dydt(t + dt / int2, u1, f1)

    for i in range(m):
        u2[i] = y0[i] + dt * f1[i] / int2
    dydt(t + dt / int2, u2, f2)

    for i in range(m):
        u3[i] = y0[i] + dt * f2[i]
    dydt(t + dt, u3, f3)

    for i in range(m):
        y[i] = y0[i] + dt * (f0[i] + int2 * f1[i] + int2 * f2[i] + f3[i]) / int6

File: Matlab/rk4.m


function y  = rk4 (dydt, tspan, y0, n)

  y0 = y0(:);

  m = size ( y0, 1 );
  t = zeros ( n + 1, 1 );
  y_temp = zeros ( n + 1, m );

  tfirst = tspan(1);
  tlast = tspan(2);
  dt = ( tlast - tfirst ) / n;

  t(1,1) = tspan(1);
  y_temp(1,:) = y0(:);
  
  for i = 1 : n
    f1 = dydt( y_temp(i,:) );
    f2 = dydt( y_temp(i,:) + dt * f1 / 2.0 );
    f3 = dydt( y_temp(i,:) + dt * f2 / 2.0 );
    f4 = dydt( y_temp(i,:) + dt * f3 );

    t(i+1,1) = t(i,1) + dt;
    y_temp(i+1,:) = y_temp(i,:) + dt * ( f1 + 2.0 * f2 + 2.0 * f3 + f4 ) / 6.0;       
  end
  y = y_temp';
  
  return
end